### PR TITLE
Fix Measure Tool behaviour for geographic coordinate systems

### PR DESCRIPTION
**Pull request by jdugge**
## Description
Selecting "Map Units" in the Measure Tool with a geographic CRS should activate degrees. However, the Measure Tool currently uses meters in this case. This pull request fixes this behaviour.

Also, when measuring distances in degrees in a geographic coordinates system, Cartesian calculations should be used (as explained in the Info box). However, the Measure Tool currently uses distances on the ellipsoid, which are converted to degrees at the equator. Changing from degrees to other units and back to degrees correctly recalculates the distances. This pull request fixes this behaviour.

This pull request also adds centimeter and millimeter based units to Project Properties and Map Tool Options (they are currently only available in the Measure Tool).

## Checklist

- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by nyalldawson** [comment]
Nice work jdugge! 

Could you rebase your branch on current master and then force push, to avoid all the extra commits here? It'll make reviewing easier.

**Comment by jdugge** [comment]
Thanks Nyall!
I've rebased my branch and force pushed now.

File: src/app/qgsmeasuredialog.h [hunk 1]
```diff
@@ -98,6 +98,15 @@ class APP_EXPORT QgsMeasureDialog : public QDialog, private Ui::QgsMeasureBase
     //! indicates whether we're measuring distances or areas
     bool mMeasureArea = false;
 
+    //! Indicates whether the user chose "Map units" instead of directly selecting a unit
+    bool useMapUnits = false;
```
Comment (nyalldawson) [inline src/app/qgsmeasuredialog.h]: QGIS code convention is that member variables all have a "m" prefix, so this should be `mUseMapUnits` (and `mMapDistanceUnits`, etc)

File: python/core/auto_generated/qgsdistancearea.sip.in [hunk 2]
```diff
@@ -189,23 +189,25 @@ Measures the perimeter of a polygon geometry.
 .. versionadded:: 2.12
 %End
 
-    double measureLine( const QVector<QgsPointXY> &points ) const;
+    double measureLine( const QVector<QgsPointXY> &points, bool forceCartesian = false ) const;
```
Comment (nyalldawson) [inline python/core/auto_generated/qgsdistancearea.sip.in]: I think we can avoid these new "forceCartesian" arguments here -- what we could do instead is from qgsmeasuredialog.cpp call `mDa.setEllipsoid( QString() )`. That will disable ellipsoidal calculation. And then to re-enable the ellipsoidal calculation you'd call `mDa.setEllipsoid( QgsProject::instance()->ellipsoid() )`. 

I'd rather avoid the extra argument here and keep the QgsDistanceArea simpler (I'm sure you'll agree it's already quite complex!)

**Comment by nyalldawson** [comment]
Looking good- and a valuable fix to have in place. Can you address the above comments and I'll merge?

File: src/app/qgsmeasuredialog.h (hunk 1 above)
Comment (jdugge) [reply src/app/qgsmeasuredialog.h]: Done

File: python/core/auto_generated/qgsdistancearea.sip.in (hunk 2 above)
Comment (jdugge) [reply python/core/auto_generated/qgsdistancearea.sip.in]: Done (although I had to use `mDa.setEllipsoid( GEO_NONE )` for this to work)

**Comment by jdugge** [comment]
I'm not sure why the Travis CI build fails: the tests that time out work fine on my system. Do you have an idea what could be causing the timeouts?

**Comment by nyalldawson** [comment]
Perfect -- thank you!

**Pull request merged**